spec/statement.dd

```diff
@@ -808,60 +814,180 @@ $(GNAME OpApplyParameter):
+    $(BEST_PRACTICE Annotate delegate parameters to apply functions with `scope` when possible.)
+
+    $(PANEL
+        The `scope` storage class on the `body` parameter means that the delegate does
+        not escape the scope of the apply function (e.g. assigning `body` to a
+        global variable). If it cannot be statically guaranteed that `body` does not escape, a closure may
+        be allocated for it on the heap instead of the stack.
     )
 
-    $(P $(B Important:) If $(D opApply) catches any exceptions, ensure that those
-        exceptions did not originate from the delegate passed to $(D opApply). The user would expect
-        exceptions thrown from a `foreach` body to both terminate the loop, and propagate outside
+    $(NOTE If an apply function catches any exceptions, ensure that those
+        exceptions did not originate from the delegate. The user would expect
+        exceptions thrown from a `foreach` body to both terminate the loop and propagate outside
         the `foreach` body.
      )
 
 $(H4 $(LNAME2 template-op-apply, Template `opApply`))
 
-    $(P $(D opApply) can also be a templated function,
-        which will infer the types of parameters based on the $(I ForeachStatement).
-        For example:)
+    $(P `opApply` and `opApplyReverse` can also be function templates,
+        which can optionally infer the types of parameters based on the $(I ForeachStatement).
+    )
+
+    $(NOTE An apply function template cannot infer `foreach` variable types.)
 
         $(SPEC_RUNNABLE_EXAMPLE_RUN
         --------------
         struct S
         {
-            import std.traits : ParameterTypeTuple;  // introspection template
-            import std.stdio;
-
-            int opApply(Dg)(scope Dg dg)
-            if (ParameterTypeTuple!Dg.length == 2) // foreach with 2 parameters
+            int opApply(Body)(scope Body body)
             {
-                writeln(2);
-                return 0;
-            }
-
-            int opApply(Dg)(scope Dg dg)
-            if (ParameterTypeTuple!Dg.length == 3) // foreach with 3 parameters
-            {
-                writeln(3);
+                pragma(msg, Body);
                 return 0;
             }
         }
 
         void main()
         {
-            foreach (int a, int b; S()) { }  // calls first opApply function
-            foreach (int a, int b, float c; S()) { }  // calls second opApply function
+            foreach (int a, int b; S()) { } // int delegate(ref int, ref int) pure nothrow @nogc @safe
+            foreach (bool b, string s; S()) { } // int delegate(ref bool, ref string) pure nothrow @nogc @safe
+            //foreach (x; S()) { } // Error: cannot infer type for `foreach` variable `x`, perhaps set it explicitly
         }
         --------------
         )
 
+$(H4 $(LNAME2 template-instance-op-apply, `opApply` as an Alias to an Explicit Function Template Instance))
```

Comment:
This might be a better subheading:
```suggestion
$(H4 $(LNAME2 template-instance-op-apply, `opApply` with Attribute Inference))
```

Reply:
Agreed. This is a “what it intends to do” vs. “what it technically is” ― and usually, “what it intends” is better. When I wrote this section, I focused on getting the technicalities right.

The introductory sentence to the subsection does not read weird, so I can just take your suggestion. 👍🏻